source/connection_manager.c

```diff
@@ -0,0 +1,876 @@
+ *
+ */
+struct aws_http_connection_manager {
+    struct aws_allocator *allocator;
+
+    /*
+     * A union of external downstream dependencies (primarily global http API functions) and
+     * internal implementation references.  Selectively overridden by tests in order to
+     * enable strong coverage of internal implementation details.
+     */
+    const struct aws_http_connection_manager_function_table *function_table;
+
+    /*
+     * Controls access to all mutable state on the connection manager
+     */
+    struct aws_mutex lock;
+
+    /*
+     * A manager can be in one of two states, READY or SHUTTING_DOWN.  The state transition
+     * takes place when ref_count drops to zero.
+     */
+    enum aws_http_connection_manager_state_type state;
+
+    /*
+     * The set of all available, ready-to-be-used connections
+     */
+    struct aws_array_list connections;
+
+    /*
+     * The set of all incomplete connection acquisition requests
+     */
+    struct aws_linked_list pending_acquisitions;
+
+    /*
+     * The number of all incomplete connection acquisition requests.  So
+     * that we don't have compute the size of a linked list every time.
+     */
+    size_t pending_acquisition_count;
+
+    /*
+     * The number of pending new connection requests we have outstanding to the http
+     * layer.
+     */
+    size_t pending_connects_count;
+
+    /*
+     * The number of connections currently being used by external users.
+     */
+    size_t vended_connection_count;
+
+    /*
+     * Always equal to # of connection shutdown callbacks not yet invoked
+     * or equivalently:
+     *
+     * # of connections ever created by the manager - # shutdown callbacks received
+     */
+    size_t open_connection_count;
+
+    /*
+     * All the options needed to create an http connection
+     */
+    struct aws_client_bootstrap *bootstrap;
+    size_t initial_window_size;
+    struct aws_socket_options socket_options;
+    struct aws_tls_connection_options *tls_connection_options;
+    struct aws_string *host;
+    uint16_t port;
+
+    /*
+     * The maximum number of connections this manager should ever have at once.
+     */
+    size_t max_connections;
+
+    /*
+     * Lifecycle tracking for the connection manager.  Starts at 1.
+     *
+     * Once this drops to zero, the manager state transitions to shutting down
+     *
+     * The manager is deleted when all other tracking counters have returned to zero.
+     *
+     * We don't use an atomic here because the shutdown phase wants to check many different
+     * values.  You could argue that we could use a sum of everything, but we still need the
+     * individual values for proper behavior and error checking during the ready state.  Also,
+     * a hybrid atomic/lock solution felt excessively complicated and delicate.
+     */
+    size_t external_ref_count;
+};
+
+/*
+ * Hard Requirement: Manager's lock must held somewhere in the call stack
+ */
```

Comment:
just an idea: have a bool that tracks whether it's locked, and you can assert on that in your functions.
I guess you could never assert that it's unlocked, but it would be safe to assert that it was locked

```
static void s_lock(manager) {
  mutex_lock(manager->lock);
  manager->is_locked = true;
}

Reply:
Unfortunately, that wouldn't reliably verify that your thread is the one who took the lock though.